jetbrains-core/src/software/aws/toolkits/jetbrains/core/credentials/ConnectionSettingsMenuBuilder.kt

```diff
@@ -0,0 +1,193 @@
+// Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.core.credentials
+
+import com.intellij.openapi.actionSystem.ActionManager
+import com.intellij.openapi.actionSystem.ActionPlaces
+import com.intellij.openapi.actionSystem.AnAction
+import com.intellij.openapi.actionSystem.AnActionEvent
+import com.intellij.openapi.actionSystem.DataContext
+import com.intellij.openapi.actionSystem.DefaultActionGroup
+import com.intellij.openapi.actionSystem.Separator
+import com.intellij.openapi.actionSystem.ToggleAction
+import com.intellij.openapi.project.DumbAware
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.ui.popup.JBPopupFactory
+import com.intellij.openapi.ui.popup.ListPopup
+import software.aws.toolkits.core.credentials.CredentialIdentifier
+import software.aws.toolkits.core.region.AwsRegion
+import software.aws.toolkits.jetbrains.core.region.AwsRegionProvider
+import software.aws.toolkits.jetbrains.core.utils.buildList
+import software.aws.toolkits.resources.message
+
+class ConnectionSettingsMenuBuilder private constructor() {
+    private data class RegionSelectionSettings(val currentSelection: AwsRegion?, val onChange: (AwsRegion) -> Unit)
+    private data class CredentialsSelectionSettings(val currentSelection: CredentialIdentifier?, val onChange: (CredentialIdentifier) -> Unit)
+
+    private var regionSelectionSettings: RegionSelectionSettings? = null
+    private var credentialsSelectionSettings: CredentialsSelectionSettings? = null
+    private var accountSettingsManager: AwsConnectionManager? = null
+
+    fun withRegions(currentSelection: AwsRegion?, onChange: (AwsRegion) -> Unit): ConnectionSettingsMenuBuilder {
```

Comment:
nit: `withRegions()` and `withRegionSelected(currentSelection)`? not sure how you'd clarify that it'd be either/or

Reply:
Not sure there is a great way to "solve" that without interface / type safe builders which is a lot of boilerplate for this.